### PR TITLE
[Core][Performance] Add XGrammar support for guided decoding and set it as default

### PR DESCRIPTION
**Pull request by aarnphm**
Add initial support for [XGrammar](https://github.com/mlc-ai/xgrammar) for V0 and makes it the default for grammar and json usage. Written in collaboration with @mgoin 

I'm using the benchmark scripts from #10557

Results for using XGrammar as backend:

```text
Throughput: 0.94 requests/s, 1022.46 total tokens/s, 480.27 output tokens/s Correct rate is 100.0 %
First token latency(msecs):
count      10.000000
mean     4552.206317
std       734.671745
min      3289.774953
25%      3864.269087
50%      5102.686635
75%      5102.717258
max      5114.346570
dtype: float64
Next token latency(msecs):
count    10.000000
mean     11.906452
std       1.409063
min      10.831970
25%      10.837367
50%      10.854235
75%      13.227200
max      14.325024
dtype: float64
```

Comparing to outlines

```text
Throughput: 0.22 requests/s, 241.22 total tokens/s, 113.31 output tokens/s Correct rate is 100.0 %
First token latency(msecs):
count       10.000000
mean     38533.083248
std         35.807892
min      38491.813741
25%      38491.826321
50%      38556.601226
75%      38556.628519
max      38568.547848
dtype: float64
Next token latency(msecs):
count    10.000000
mean     12.955556
std       0.042220
min      12.901755
25%      12.914099
50%      12.953058
75%      12.996646
max      13.003127
dtype: float64
```

NOTE: Running on A100 80GB, with Llama 3.2 3B with chunked prefill enable and JSON grammar

### CODE REVIEW

**Review by Ubospica — COMMENTED** [review]
Thanks for your contribution to integrating XGrammar into vLLM! It overall looks good, but there are some minor points to enhance parallelism.

File: vllm/model_executor/guided_decoding/xgrammar_decoding.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+if TYPE_CHECKING:
+    from transformers import PreTrainedTokenizer
+
+    from vllm.config import ModelConfig
+    from vllm.sampling_params import GuidedDecodingParams
+
+
+# TODO: passing batch size to max threads here
+def get_local_xgrammar_guided_decoding_logits_processor(
```
[the remark this answers is not in the document]
Comment (mgoin) [reply vllm/model_executor/guided_decoding/xgrammar_decoding.py:25]: > Besides, GrammarCompiler and TokenizerInfo should be initialized once for a model (if one engine serves only one model, these classes should be initialized engine-wise).

I'm looking into this right now as it really hurts TTFT and chances of caching by recreating these

[the remark this answers is not in the document]
Comment (aarnphm) [reply vllm/model_executor/guided_decoding/xgrammar_decoding.py:25]: Hmm, the reason why I'm currently having to initialize lazily is that `GrammarMatcher` is not pickable. so for multiprocessing client we need to reconstruct it after the client send the ruquest

File: vllm/model_executor/guided_decoding/xgrammar_decoding.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+            self.prefilled = True
+        else:
+            for i, matcher in enumerate(self.matchers):
+                if not matcher.is_terminated():
+                    sampled_token = input_ids[-1]
+                    assert self.matchers[i].accept_token(sampled_token)
+
+        for i, matcher in enumerate(self.matchers):
+            if not matcher.is_terminated():
+                matcher.fill_next_token_bitmask(self.token_bitmask, i)
```
[the remark this answers is not in the document]
Comment (aarnphm) [reply vllm/model_executor/guided_decoding/xgrammar_decoding.py:165]: I think in v0 logit_processor is row blocking, so i'm not entirely sure how to parallelize it atm.

File: vllm/model_executor/guided_decoding/__init__.py [hunk 3]
```diff
@@ -1,12 +1,18 @@
...
+    from vllm.sampling_params import GuidedDecodingParams
 
 
 async def get_guided_decoding_logits_processor(
-        guided_params: GuidedDecodingParams,
-        tokenizer) -> Optional[LogitsProcessor]:
+        guided_params: GuidedDecodingParams, tokenizer: PreTrainedTokenizer,
+        model_config: ModelConfig) -> LogitsProcessor | None:
     # CFG grammar not supported by LMFE, so we use outlines instead
     if guided_params.backend == 'outlines' or guided_params.grammar:
```
[the remark this answers is not in the document]
Comment (aarnphm) [reply vllm/model_executor/guided_decoding/__init__.py:17]: i see, I will add this difference into the docs

[the remark this answers is not in the document]
Comment (aarnphm) [reply vllm/model_executor/guided_decoding/__init__.py]: I think we should just remove the grammar check here.

If user send grammar they should also specify the backend (probably better to document the cartesian product of the combinations)

**Comment by mgoin** [comment]
Updated this PR with caches for the tokenizer data and the grammar compiler to avoid constructing these data structures for each request. It isn't pretty but it boosts throughput by about 1.4x

Command: 
```
python benchmark_guided.py --model meta-llama/Llama-3.1-8B-Instruct --dataset xgrammar_bench --async-engine --output-len 512 --num-prompts 20 --enable-chunked-prefill --guided-decoding-ratio 1
```

Before:
```
Throughput: 1.46 requests/s, 1189.12 total tokens/s, 748.00 output tokens/s Correct rate is 95.0 % 
First token latency(msecs):
count      20.000000
mean     7180.142369
std      1212.973158
min      4644.173431
25%      7012.610644
50%      7578.541221
75%      8079.524654
max      8092.886029
dtype: float64
Next token latency(msecs):
count    20.000000
mean     12.662371
std       2.336552
min      10.942158
25%      10.942283
50%      11.864077
75%      12.990130
max      17.550802
dtype: float64
```

After:
```
Throughput: 2.12 requests/s, 1726.67 total tokens/s, 1086.13 output tokens/s Correct rate is 95.0 % 
First token latency(msecs):
count      20.000000
mean     3254.682581
std       290.516334
min      2869.083916
25%      2869.120228
50%      3449.280638
75%      3477.460549
max      3477.504314
dtype: float64
Next token latency(msecs):
count    20.000000
mean     12.054585
std       0.550868
min      11.643879
25%      11.643967
50%      11.674903
75%      12.786106
max      12.786302
dtype: float64
```

**Comment by mgoin** [comment]
@Ubospica do you know when XGrammar can support regex? This would help with covering existing use cases

**Comment by joennlae** [comment]
@mgoin I added a pull request yesterday that adds some simple regex pattern + integer ranges support:

https://github.com/mlc-ai/xgrammar/pull/106

File: vllm/entrypoints/llm.py [hunk 4]
```diff
@@ -1037,7 +1038,8 @@ def _validate_and_add_requests(
         for i, prompt in enumerate(prompts):
             self._add_request(
                 prompt,
-                params[i] if isinstance(params, Sequence) else params,
+                params[i]
+                if isinstance(params, Sequence) else copy.copy(params),
```
Comment (simon-mo) [inline vllm/entrypoints/llm.py:1042]: This copy is likely going to lower the throughput, any chance that we can remove it?

**Review by simon-mo — DISMISSED** [review]

File: vllm/entrypoints/llm.py (hunk 4 above)
Comment (mgoin) [reply vllm/entrypoints/llm.py:1042]: Okay, I can move this into the function where we get the guided decoding logit processor

**Comment by mgoin** [comment]
Thanks for review @simon-mo I moved the copy into a specific `if sampling_params.guided_decoding is not None` case - ready for re-review

File: vllm/model_executor/guided_decoding/xgrammar_decoding.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (aarnphm) [reply vllm/model_executor/guided_decoding/xgrammar_decoding.py:25]: sync offline, xgrammar will provide a serialisation for compiled grammar as json string, in which we can initialized the grammar server-side

File: vllm/model_executor/guided_decoding/xgrammar_decoding.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (aarnphm) [reply vllm/model_executor/guided_decoding/xgrammar_decoding.py:236]: Sync offline, will look into this for v1.

**Review by simon-mo — APPROVED** [review]

**Comment by hmellor** [comment]
The new dependency in this PR appears to have broken installation on ARM

```
8.373 ERROR: Could not find a version that satisfies the requirement xgrammar (from versions: none)
8.419 ERROR: No matching distribution found for xgrammar
------
Dockerfile.arm:37
--------------------
  36 |     
  37 | >>> RUN --mount=type=cache,target=/root/.cache/pip \
  38 | >>>     --mount=type=bind,src=requirements-common.txt,target=requirements-common.txt \
  39 | >>>     --mount=type=bind,src=requirements-cpu.txt,target=requirements-cpu.txt \
  40 | >>>     pip install -v -r requirements-cpu.txt
  41 |     
--------------------
ERROR: failed to solve: process "/bin/sh -c pip install -v -r requirements-cpu.txt" did not complete successfully: exit code: 1
```

**Comment by mgoin** [comment]
Thanks for reporting @hmellor indeed it seems there isn't a manylinux arm wheel available https://pypi.org/project/xgrammar/#files

I'll work on a patch fix

**Comment by stefanobranco** [comment]
Obviously super cool to see new integrations, but it does seem a bit hasty to me to immediately change the default? The implementation with outlines core should be able to close the gap after all, and this one does not support regex yet. Or is xgrammar just objectively better?

**Comment by joennlae** [comment]
I second this opinion. Currently, the same behaviour cannot be expected from 'grammar`. I added a simple PR with some rudimentary regex + integer range support (https://github.com/mlc-ai/xgrammar/pull/106). 

I can attest that it is much faster, especially if one uses dynamic schemas. However, we should use `outlines` as the default, as it supports more cases for now, and the change is not breaking for many.

I introduced it as an option in my closed PR (#10803). But I forgot it when I discussed it with @mgoin.

**Comment by mgoin** [comment]
Hi @stefanobranco and @joennlae thanks for raising your concern. Our primary concern is immediately improving structured output performance where it is easy to do so while maintaining the same behavior. With xgrammar as the default, we still fallback to outlines in several cases covered here https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/guided_decoding/__init__.py#L18-L48

Please let me know if a case isn't being accounted for that is affecting your usage. We do not want to change external behavior. We have several integration tests that I have been using to create these rules, but more test points are certainly welcome!

We have several fast-followup items to reduce the special cases around using xgrammar and improving performance even further in V0. We are also working on enabling outlines>1.0 support with the devs of that project. Then of course we will enable the usage of structured output in V1.

I hope this is helpful context and we will work on making a public roadmap for longer term goals. Please join the #feat-structured-output channel in slack if you want to have more direct discussion with the people working on this.

**Comment by Ubospica** [comment]
Thanks @stefanobranco and @joennlae for pointing out your concern, and @mgoin for your explanation. 

Currently, XGrammar's plan is to provide full support for JSON Schema (cover all common corner cases), and support regex. These are expected to be finished this week. I think after that from a functional perspective XGrammar will cover all cases that outlines can support. 

For grammar the main difference is the format is different, but I think it should be easy to convert. We can provide a conversion utility first.

From a performance perspective, for json schema/regex, XGrammar is much faster in compiling even for Outlines v0.1 (30ms vs 1s). For CFG, the mask generation time is significantly enhanced (30us vs 5ms). That should be important in serving usecases.

**Comment by ktrapeznikov** [comment]
will this support models that use mistral tokenizers?

**Comment by robcaulk** [comment]
@joennlae Pointed out correctly that changing the default value from outlines to xgrammar was a breaking change. This should have been highlighted in the release notes as a breaking change.

@mgoin you had the foresight to avoid changing behavior, but unfortunately, this change did change the behavior. The issue now is that the quality of output from xgrammar is not as high. It does not conform to Literal definitions in the schema. Outlines does.

This broke quite a bit of our pipeline - as we require Literals. We will define outlines explicitly now to avoid the shortcoming of xgrammar, but I highly recommend to the maintainers (@simon-mo) that any breaking changes be properly highlighted in release notes in the future.
